src/wagtail_factories/factories.py

```diff
@@ -144,3 +146,26 @@ class Meta:
 
     title = "A document"
     file = factory.django.FileField()
+
+
+class RedirectFactory(DjangoModelFactory):
+    old_path = factory.Faker("slug")
+    site = factory.SubFactory(SiteFactory)
+    redirect_page = factory.SubFactory(PageFactory)
+    redirect_link = factory.Faker("url")
+    is_permanent = factory.Faker("boolean")
+
+    class Meta:
+        model = Redirect
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
+        """
+        Override _create() to ensure that Redirect.clean() is run
+        in order to normalise `old_path`.
+        @see https://github.com/wagtail/wagtail/blob/main/wagtail/contrib/redirects/models.py#L191
+        """
+        obj = model_class(*args, **kwargs)
+        obj.clean()
```

Comment:
Issue: I don't think we should be calling `clean` here, as it may clobber the values provided by a user. There are scenarios in which you may want to provide invalid data to a model instance — for example, if you want to test the `clean` method itself. I think our approach should be to make a best effort to generate example data that is both valid, and matches the typical usage of the object — but give the user the ability to create objects with data however they want, if they need to.

I would _love_ to see a custom faker provider developed for `old_path`. We could use [uri_path](https://faker.readthedocs.io/en/stable/providers/faker.providers.internet.html#faker.providers.internet.Provider.uri_path) as a starting point (just using slugs is a little deficient). `old_path` also accepts query parameters, for which we could use [pydict](https://faker.readthedocs.io/en/stable/providers/faker.providers.python.html#faker.providers.python.Provider.pydict) as a source. An API using [traits](https://factoryboy.readthedocs.io/en/stable/introduction.html#altering-a-factory-s-behavior-parameters-and-traits) to enable flags like `has_query_params` would be the icing on the cake.

Reply:
A bit more context on whether or not we should call `clean`:
- Django does not call `full_clean` in `save` by default (https://docs.djangoproject.com/en/5.0/ref/models/instances/#django.db.models.Model.full_clean)
- Wagtail _does_ for `Page` models, but it is optional, and exists to facilitate restoring revisions (https://github.com/wagtail/wagtail/blob/main/wagtail/models/__init__.py#L1495)